File: Julia-files/05_writing_functions.jl
using Statistics

# 1. Cube Function ----------------------------------------

cube(x) = x.^3

println(cube(1:10))

m = reshape(1:8, 2, 4)   # Julia fills column-major like R
println(cube(m))

m2 = reshape(cube(1:8), 2, 4)
println(m2)

a = reshape(1:24, 2, 3, 4)
println(cube(a))

println(typeof(cube))


# 2. Robust Loss Functions ---------------------------------

psi1(x) = x.^2 .> 1 ? 2 .* abs.(x) .- 1 : x.^2

psi2(x; c=1) = x.^2 .> c^2 ? 2c .* abs.(x) .- c^2 : x.^2

function psi3(x; c=1)
    @assert c > 0 "c must be positive"
    @assert length(c) == 1 "c must be scalar"
    return x.^2 .> c^2 ? 2c .* abs.(x) .- c^2 : x.^2
end

z = [-3, -1, 0, 1, 3]

println(psi1.(z) == psi2.(z, c=1))
println(psi2.(z, c=1) == psi2.(z))
println(psi2.(z, c=2) == psi2.(z, c=2))

println(psi2.(z, c=[1,1,1,10,1]))


# 3. Scoping Example ----------------------------------------

global x = 7
y = ["A","C","G","T","U"]

function adder(y)
    global x
    x = x + y
    return x
end

println(adder(1))


# 4. Circle Area Function ------------------------------------

circle_area(r) = π .* r.^2

println(circle_area([1,2,3]))

realpi = π
π = 3
println(circle_area([1,2,3]))

π = realpi
println(circle_area([1,2,3]))


# 5. Recursive Fibonacci ---------------------------------------

function fib(n)
    n == 0 || n == 1 ? 1 : fib(n-1) + fib(n-2)
end

println(fib(6))


# 6. Predicting From Power-Law Model ---------------------------

function predict_plm(model::Dict, newdata)
    @assert haskey(model, "a") && haskey(model, "y0")
    a = model["a"]
    y0 = model["y0"]
    return y0 .* newdata.^a
end

mod = Dict("a"=>0.15, "y0"=>6611)
println(predict_plm(mod, [10, 100, 1000]))

File: Python-files/05_writing_functions.py
import numpy as np

# Define the cube function
def cube(x):
    return np.power(x, 3)

# cube(1:10)
print(cube(np.arange(1, 11)))
# Output: [   1    8   27   64  125  216  343  512  729 1000]

# cube(matrix(1:8, 2, 4))
m = np.arange(1, 9).reshape(2, 4, order='F')  # 'F' means fill column-wise like R
print(cube(m))
# Output:
# [[  1  27 125 343]
#  [  8  64 216 512]]

# matrix(cube(1:8), 2, 4)
m2 = np.array(cube(np.arange(1, 9))).reshape(2, 4, order='F')
print(m2)
# Output same as above

# cube(array(1:24, c(2, 3, 4)))
a = np.arange(1, 25).reshape(2, 3, 4, order='F')
print(cube(a))  # cubes each element in the 3D array

# mode(cube)
print(type(cube))  # <class 'function'>

# Robust Loss functions
import numpy as np

def psi1(x):
    x = np.asarray(x)
    return np.where(x**2 > 1, 2 * np.abs(x) - 1, x**2)

def psi2(x, c=1):
    x = np.asarray(x)
    return np.where(x**2 > c**2, 2 * c * np.abs(x) - c**2, x**2)

def psi3(x, c=1):
    if not np.isscalar(c) or c <= 0:
        raise ValueError("c must be a single positive number")
    x = np.asarray(x)
    return np.where(x**2 > c**2, 2 * c * np.abs(x) - c**2, x**2)

# function scoping and constants
x = 7
y = ["A", "C", "G", "T", "U"]

def adder(y):
    global x   # This will cause UnboundLocalError in Python (if global declaration is ignored))
    x = x + y  
    return x

# Circle Area function
def circle_area(r):
    """
    Calculate the area of a circle given radius r.
    Supports scalar or iterable inputs.
    """
    r = np.array(r)
    return np.pi * r**2

print(circle_area([1, 2, 3]))  # Output: [3.14, 12.57, 28.27]

# Recursive Fibonacci function
def fib(n):
    """
    Recursive Fibonacci function.
    Returns 1 for n=0 or n=1, else fib(n-1) + fib(n-2).
    """
    if n == 0 or n == 1:
        return 1
    else:
        return fib(n - 1) + fib(n - 2)

# Predict from a fitted model
def predict_plm(model, newdata):
    """
    Predict response values from a power-law scaling model.
    model: dict with keys 'a' and 'y0'
    newdata: numeric or array of values to predict on
    """
    assert 'a' in model and 'y0' in model, "Model must contain 'a' and 'y0'"
    a, y0 = model['a'], model['y0']
    assert isinstance(a, (int, float)), "a must be numeric"
    assert isinstance(y0, (int, float)), "y0 must be numeric"
    newdata = np.array(newdata)
    return y0 * newdata**a


